pkg/controller/containerjfr/containerjfr_controller_test.go

```diff
@@ -0,0 +1,134 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package containerjfr_test
+
+import (
+	"context"
+	"net/http"
+
+	certv1 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	certMeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/operator-framework/operator-sdk/pkg/log/zap"
+	rhjmcv1beta1 "github.com/rh-jmc-team/container-jfr-operator/pkg/apis/rhjmc/v1beta1"
+	"github.com/rh-jmc-team/container-jfr-operator/pkg/controller/containerjfr"
+	"github.com/rh-jmc-team/container-jfr-operator/test"
+)
+
+var _ = Describe("ContainerjfrController", func() {
+	var (
+		objs       []runtime.Object
+		handlers   []http.HandlerFunc
+		server     *test.ContainerJFRServer
+		client     client.Client
+		controller *containerjfr.ReconcileContainerJFR
+	)
+
+	JustBeforeEach(func() {
+		logf.SetLogger(zap.Logger())
+		s := test.NewTestScheme()
+
+		client = fake.NewFakeClientWithScheme(s, objs...)
+		server = test.NewServer(client, handlers)
+		controller = &containerjfr.ReconcileContainerJFR{
+			Client:        client,
+			Scheme:        s,
+			ReconcilerTLS: test.NewTestReconcilerTLS(client),
+		}
+	})
+
+	JustAfterEach(func() {
+		server.VerifyRequestsReceived(handlers)
+		server.Close()
+	})
+
+	BeforeEach(func() {
+		objs = []runtime.Object{
+			test.NewContainerJFR(), test.NewCACert(), test.NewTargetPod(),
+			test.NewContainerJFRService(), test.NewJMXAuthSecret(),
+		}
+	})
+
+	AfterEach(func() {
+		objs = nil
+	})
+
+	Describe("reconciling a request", func() {
+		Context("after Containerjfr already reconciled successfully", func() {
+			It("should be idempotent", func() {
+				req := reconcile.Request{NamespacedName: types.NamespacedName{Name: "containerjfr", Namespace: "default"}}
+				result, err := controller.Reconcile(req)
+				Expect(err).ToNot(HaveOccurred())
+				//Expect(result).To(Equal(reconcile.Result{}))
+
+				obj := &rhjmcv1beta1.ContainerJFR{}
+				err = client.Get(context.Background(), req.NamespacedName, obj)
+				Expect(err).ToNot(HaveOccurred())
+
+				caCert := &certv1.Certificate{}
+				err = client.Get(context.Background(), types.NamespacedName{Name: "containerjfr-ca", Namespace: "default"}, caCert)
+				Expect(err).ToNot(HaveOccurred())
+				caCert.Status.Conditions = append(caCert.Status.Conditions, certv1.CertificateCondition{
+					Type:   certv1.CertificateConditionReady,
+					Status: certMeta.ConditionTrue,
+				})
+				err = client.Status().Update(context.Background(), caCert)
+				Expect(err).ToNot(HaveOccurred())
```

Comment:
You'll have to do this as well for the other Certificates created by the operator: `containerjfr`, `containerjfr-grafana`.

Additionally, you'll have to create the associated Secrets as well, as cert-manager would. The secret names should match each Certificate's spec.SecretName property. For these tests, I think it will be sufficient to fill the Secrets with dummy values under the `corev1.TLSCertKey`, `corev1.TLSPrivateKeyKey` and `resource_definitions.CAKey` keys.